src/test/java/net/openhft/chronicle/network/UberHandlerTest.java

```diff
@@ -241,88 +255,147 @@ public void testBusyWritingHandlersAreCalledFirstInRoundRobin() throws IOExcepti
+    }
 
-        try (MyClusterContext acceptorCtx = clusterContext(acceptorHost, initiatorHost);
-             MyClusterContext initiatorCtx = clusterContext(initiatorHost, acceptorHost)) {
+    @Test
+    public void rejectedOnReadHandlersAreRemoveFromReadAndWrite() {
+        try (final UberHandlerTestHarness testHarness = new UberHandlerTestHarness()) {
+            testHarness.registerSubHandler(new WritableRejectingSubHandler());
+            expectException("Rejected in onRead");
+            REJECTING_SUB_HANDLER_SHOULD_REJECT.set(true);
+            testHarness.sendMessageToCurrentHandler();
+            expectException("handler == null, check that the Csp/Cid has been sent");
+            testHarness.sendMessageToCurrentHandler();
+            testHarness.callProcess();
+            assertFalse(REJECTED_HANDLER_ONWRITE_CALLED.get());
+            assertFalse(REJECTED_HANDLER_ONREAD_CALLED.get());
+        }
+    }
 
-            acceptorCtx.cluster().start(acceptorHost.hostId());
-            initiatorCtx.cluster().start(initiatorHost.hostId());
+    @Test
+    public void rejectedOnWriteHandlersAreRemoveFromReadAndWrite() {
+        try (final UberHandlerTestHarness testHarness = new UberHandlerTestHarness()) {
+            testHarness.registerSubHandler(new WritableRejectingSubHandler());
+            expectException("Rejected in onWrite");
+            REJECTING_SUB_HANDLER_SHOULD_REJECT.set(true);
+            testHarness.callProcess();
+            expectException("handler == null, check that the Csp/Cid has been sent");
+            testHarness.sendMessageToCurrentHandler();
+            testHarness.callProcess();
+            assertFalse(REJECTED_HANDLER_ONWRITE_CALLED.get());
+            assertFalse(REJECTED_HANDLER_ONREAD_CALLED.get());
+        }
+    }
 
-            initiatorCtx.connectionManager(acceptorHost.hostId()).addListener((nc, isConnected) -> {
-                if (isConnected) {
-                    // This handler should reject in onInitialize
-                    expectException("Rejected in onInitialize");
-                    nc.wireOutPublisher().publish(w ->
-                            w.writeDocument(true, d -> sendHandler(d, FIRST_CID, new RejectingOnInitializeHandler())));
-                    expectException("handler == null, check that the Csp/Cid has been sent");
-                    nc.wireOutPublisher().publish(w ->
-                            w.writeDocument(false, d -> sendMessageToHandler(d, FIRST_CID)));
-
-                    // This handler should reject in onRead
-                    nc.wireOutPublisher().publish(w ->
-                            w.writeDocument(true, d -> sendHandler(d, FIRST_CID + 1, new RejectingOnReadHandler())));
-                    expectException("Rejected in onRead");
-                    nc.wireOutPublisher().publish(w ->
-                            w.writeDocument(false, d -> sendMessageToHandler(d, FIRST_CID + 1)));
-                }
-            });
```

Comment:
moved this test to use the test harness (could probably do more but will stop there)